# 1.1 Algorithms

### Euclid's Algorithm

    E1. r = m % n
    E2. if r == 0: return n
    E3. m, n = n, r

<hr>
Working through the example for m = 119, n = 544

    E1. r = 119
    E2. r != 0                 # step skipped
    E3. m = 544, n = 119       # if n > m: m and n exchanged in first pass

    E1. r = 68
    E3. m = 119, n = 68

    E1. r = 51
    E3. m = 68, n = 51

    E1. r = 17
    E3. m = 51, n = 17

    E1. r = 0
    E2. n = 17 = GCD

<hr>

“Assuming that the value of n is known but m is allowed to range over all positive integers, what is the average number of times, $T_n$, that step E1 of Algorithm E will be performed?”

Lets write a program to find that...


In [37]:
from math import log as ln

def countAvgSteps(n):
    count = 0
    for m in range(0, n):
        steps = 1   
        p, q = n, m
        while q != 0:
            p = p % q
            p, q = q, p
            steps += 1
        count += steps
    return float(count) / n

print('%4s  %6s  %6s' % ('n', 'avg', 'comp'))
for n in range(50, 500, 50):
    print('%4s  %6.4f  %6.4f' % (n, countAvgSteps(n), 0.9098386 + 0.803376*ln(n)))
for n in range(500, 5000, 500):
    print('%4s  %6.4f  %6.4f' % (n, countAvgSteps(n), 0.9098386 + 0.803376*ln(n)))
for n in range(5000, 50001, 5000):
    print('%4s  %6.4f  %6.4f' % (n, countAvgSteps(n), 0.9098386 + 0.803376*ln(n)))
    

   n     avg    comp
  50  4.2200  4.0527
 100  4.5600  4.6095
 150  4.7800  4.9353
 200  5.0700  5.1664
 250  5.5160  5.3456
 300  5.1600  5.4921
 350  5.5229  5.6160
 400  5.6350  5.7232
 450  5.5889  5.8179
 500  5.9040  5.9025
1000  6.4220  6.4594
1500  6.5467  6.7851
2000  6.9990  7.0162
2500  7.2832  7.1955
3000  7.0540  7.3420
3500  7.3394  7.4658
4000  7.5695  7.5731
4500  7.3689  7.6677
5000  7.7884  7.7523
10000  8.3366  8.3092
15000  8.4108  8.6349
20000  8.9083  8.8661
25000  9.1545  9.0453
30000  8.9659  9.1918
35000  9.1999  9.3156
40000  9.4858  9.4229
45000  9.2357  9.5175
50000  9.6952  9.6022


Using simple curve fitting, we see that the average number of steps is proportional log(n). In the above table, the computed values (comp column) from the curve fitted equation `0.9098386 + 0.803376*log(n)` fits the given average values reasonably well.

This strongly suggests that the average running time of the GCD algorithm is $\propto log(n)$. Finding the exact value of $T_n$ is apparently a very difficult problem (according to the text... we will find out more about that in chapter 4).

## Exercises 1.1

In [1]:
# 1.1.1

# a, b, c, d -> b, c, d, a

a, b, c, d = range(4)

print (a, b, c, d)

t = a
a = b
b = c
c = d
d = t

print (a, b, c, d)


0 1 2 3
1 2 3 0


##### 1.1.2

**Prove that m is always greater than n at the beginning of step E1,
except possibly the first time this step occurs**

The inputs m and n for algorithm E are both positive integers.

From E1, r = m % n; r is the remainder when n divides m. This implies that r < n.

In E3, we assign **m, n = n, r**, and we showed that r < n, thus n < m after the assignment in E3.

So the second time onwards, m > n every time in step E1.


##### 1.1.3

**Change algorithm E to avoid all trivial replacements (for efficiency)**

    E1. r = m % n
    E2. if r == 0: return n
    E3. m = n, n = r

**Algorithm F**

    F1. m = m % n
    F2. if m == 0: return n
    F3. n = n % m
    F4. if n == 0: return m


In [1]:
# comparing the efficiencies of algorithms E vs F

def gcdE(m, n):
    while True:
        r = m % n
        if not r:
            return n
        m, n = n, r
        
def gcdF(m, n):
    while True:
        m = m % n
        if not m:
            return n
        n = n % m
        if not n:
            return m

        
def fibo(n):
    if n < 2:
        return n
    a, b = 0, 1
    while n > 1:
        a, b = b, a + b
        n -= 1
    return b

m, n = fibo(50), fibo(49)

In [2]:
%timeit gcdE(m, n)
%timeit gcdF(m, n)

The slowest run took 12.84 times longer than the fastest. This could mean that an intermediate result is being cached.
100000 loops, best of 3: 4.18 µs per loop
100000 loops, best of 3: 3.29 µs per loop


The above result shows that gcdF is about {{'%3.0f' % (100 - 3.29 * 100 / 4.18)}}% faster than gcdE.

#### 1.1.4

**What is the GCD of 2166 and 6099?**

Using our gcdF method defind above, `gcdF(2166, 6099) == {{gcdF(2166, 6099)}}`

##### 1.1.5

**Show that the "Procedure for Reading this set of books" fails to be a genuine algorithm on 3/5 counts.**

The 5 characteristics of an algorithm are...

    - finiteness
    - definiteness
    - input
    - output
    - effectiveness
    
The "Procedure for Reading..." fails on the following :

- The procedure is infinite.
- There is no proper output (also because the procedure is infinite)
- Step 9/12/13 lack in definiteness/effectiveness. Difficult to define "mathematically inclined" precisely.

##### 1.1.6

**What is $T_5$, the average number of times step E1 is performed when n = 5?**

To find $T_5$, first we must find a way to calculate the number of steps for each value of *m*.

We notice that every *m* can be written as $m = pn + k$, where $p \ge 0$ and $ 1 \le k \le n$. Looking at algorithm E, we can see that when the first time E1 is executed, $r = k \% n$.

This implies that the value of *p* has no bearing on the number of times E1 is executed. It is sufficient to look only at all possible values of *k* (1 through 5).

| k        | successive E1 values (k, n) | times E1 is performed ($V_{nk}$) |
| -------- | ------------ | -----|
| 1 | (1, 5) (5, 1)       |    2 |
| 2 | (2, 5) (5, 2) (2, 1) |    3 |
| 3 | (3, 5) (5, 3) (3, 2) (2, 1) |    4 |
| 4 | (4, 5) (5, 4) (4, 1) |    3 |
| 5 | (5, 5)       |    1 |

$T_5$ = Average number of times E1 is performed, when n=5, is {{13/5.0}}

##### 1.1.7
**Suppose that *m* is known and *n* is allowed to range over all positive integers; let $U_m$ be the average number of times that step E1 is executed. Show that $U_m$ is well defined. Is $U_m$ in any way related to $T_m$?**

Let us first show that $T_n$ is well defined.

As we reasoned in 1.1.6, for any *n*, *m* is of the form $m = pn + k$ where $p \ge 0$ and $1 \le k \le n$. Now, we construct a table in a manner similar to 1.1.6 for each value of *k*. Since $V_{nk}$ has a definite value for each *n* and *k*, and $T_n$ is the average value of $V_{nk}$, thus $T_n$ is well defined.


Now, let us look at $U_m$.

For a fixed *m* and varying *n*, we can divide the problem into 2 cases.

- Case 1: $n \le m$
- Case 2: $n \gt m$

For case 1, let us compare this to the table in 1.1.6. There, we have a fixed *n* and varying *m* (or *k*). If we had started the algorithm with $n \ge m$, then the first pass of algorithm E will exchange *m* and *n*, and then proceed as shown in the table.

This implies that for case 1, the average number of steps $U_m = T_m - 1$.

For case 2, the first pass will exchange *m* and *n*. From that point onwards, the algorithm will proceed exactly the same as 1.1.6. In this case, $U_m = T_m + 1$.

Except some finite number of values, $n \gt m$, thus we can ignore the result from Case 1 and the result for Case 2 is the result we want.

Thus, $U_m = T_m + 1$